src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs

```diff
@@ -0,0 +1,365 @@
+                }
+                
+                if (ParamHasInOrOutAttribute(param))
+                {
+                    // recommend using ReadOnlyArray or WriteOnlyArray instead of In/Out
+                    if (isArrayType) 
+                    { 
+                        Report(WinRTRules.ArrayMarkedInOrOut, method.GetLocation(), method.Identifier, param.Identifier); 
+                    }
+                    // if not array type, stil can't use [In] or [Out]
+                    else 
+                    { 
+                        Report(WinRTRules.NonArrayMarkedInOrOut, method.GetLocation(), method.Identifier, param.Identifier); 
+                    }
+                }
+
+                if (isArrayType)
+                {
+                    bool isOutputParam = HasModifier(param, SyntaxKind.OutKeyword);
+                    // can't be both ReadOnly and WriteOnly
+                    if (hasReadOnlyArray && hasWriteOnlyArray)
+                    {
+                        Report(WinRTRules.ArrayParamMarkedBoth, method.GetLocation(), method.Identifier, param.Identifier);
+                    }
+                    // can't be both output (writeonly) and marked read only
+                    else if (hasReadOnlyArray && isOutputParam)
+                    {
+                        Report(WinRTRules.ArrayOutputParamMarkedRead, method.GetLocation(), method.Identifier, param.Identifier);
+                    }
+                    // must have some indication of ReadOnly or WriteOnly
+                    else if (!hasWriteOnlyArray && !hasReadOnlyArray && !isOutputParam) 
+                    {
+                        Report(WinRTRules.ArrayParamNotMarked, method.GetLocation(), method.Identifier, param.Identifier);
+                    }
+                }
+                // Non-array types shouldn't have attributes meant for arrays
+                else if (hasWriteOnlyArray || hasReadOnlyArray)
+                {
+                    Report(WinRTRules.NonArrayMarked, method.GetLocation(), method.Identifier, param.Identifier);
+                }
+            }
+        }
+
+        /// <summary>Make sure any namespace defined is the same as the winmd or a subnamespace of it
+        /// If component is A.B, e.g. A.B.winmd , then D.Class1 is invalid, as well as A.C.Class2
+        /// </summary>
+        /// <param name="namespace">the authored namesapce to check</param>
+        /// <param name="assemblyName">the name of the component/winmd</param>
+        /// <returns>True iff namespace is disjoint from the assembly name</returns>
+        private bool IsInvalidNamespace(INamespaceSymbol @namespace, string assemblyName)
+        {
+            var contain = @namespace;
+            while (!contain.ContainingNamespace.IsGlobalNamespace)
+            {
+                contain = contain.ContainingNamespace;
+            }
+
+            return !contain.Name.Equals(assemblyName) && !@namespace.Name.Equals(assemblyName);

```

Comment:
What if it is A.B.C.type and A.B.WinMD or something like that.

Reply:
Yes I have tests for these and have added one that more obviously tests what you are saying -- they are blocked by issue https://github.com/dotnet/roslyn/issues/50112 

As the namespace C will have "{global namespace}" as its ContainingNamespace, instead of "A.B"